light-clients/icsxx-cf-solana/src/header.rs

```diff
@@ -1,2 +1,397 @@
-super::wrap!(cf_guest_upstream::Header<PK> as Header);
-super::wrap!(impl<PK> proto for Header);
+use crate::{
+	proto,
+	solana::{
+		blockstore::{get_completed_data_ranges, get_slot_entries_in_block},
+		shred::{shred_code::ShredCode, shred_data::ShredData, *},
+	},
+};
+use alloc::vec::Vec;
+use guestchain::PubKey;
+use proto_utils::BadMessage;
+use solana_sdk::{clock::Slot, hash::Hash, signature::Signature};
+use std::{collections::BTreeSet, convert::From, marker::PhantomData};
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Header<PK> {
+	pub shreds: Vec<Shred>,
+	_phantom: PhantomData<PK>,
+}
+
+impl<PK> Header<PK> {
+	pub(crate) fn slot(&self) -> Slot {
+		assert!(!self.shreds.is_empty(), "Header must contain at least one shred");
+		self.shreds[0].slot()
+	}
+
+	pub fn hash(&self) -> Hash {
```

Comment:
```suggestion
	pub fn calculate_hash(&self) -> Hash {
```

So it’s clear that it’s an expensive function which calculates hash when called.